src/nvim/popupmnu.c

```diff
@@ -68,19 +73,62 @@ void pum_display(pumitem_T *array, int size, int selected)
   int above_row = cmdline_row;
   int redo_count = 0;
 
+  if (!pum_is_visible) {
+    // To keep the code simple, we only allow changing the
+    // draw mode when the popup menu is not being displayed
+    pum_external = pum_wants_external;
+  }
+
 redo:
+  // Pretend the pum is already there to avoid that must_redraw is set when
+  // 'cuc' is on.
+  pum_is_visible = true;
+  validate_cursor_col();
+  pum_is_visible = false;
+
+  // anchor position: the start of the completed word
+  row = curwin->w_wrow + curwin->w_winrow;
+  if (curwin->w_p_rl) {
+    col = curwin->w_wincol + curwin->w_width - curwin->w_wcol - 1;
+  } else {
+    col = curwin->w_wincol + curwin->w_wcol;
+  }
+
+  if (pum_external) {
+    Array args = ARRAY_DICT_INIT;
+    if (array_changed) {
+      Array arr = ARRAY_DICT_INIT;
+      for (i = 0; i < size; ++i) {
+        Array item = ARRAY_DICT_INIT;
+        ADD(item, STRING_OBJ(cstr_to_string((char *)array[i].pum_text)));
+        ADD(item, STRING_OBJ(cstr_to_string((char *)array[i].pum_kind)));
+        ADD(item, STRING_OBJ(cstr_to_string((char *)array[i].pum_extra)));
+        ADD(item, STRING_OBJ(cstr_to_string((char *)array[i].pum_info)));
+        ADD(arr, ARRAY_OBJ(item));
+      }
+      ADD(args, ARRAY_OBJ(arr));
+      ADD(args, INTEGER_OBJ(selected));
+      ADD(args, INTEGER_OBJ(row));
+      ADD(args, INTEGER_OBJ(col));
+      ui_event("popupmenu_show", args);
+    } else {
+      ADD(args, INTEGER_OBJ(selected));
+      ui_event("popupmenu_select", args);
+    }
+    pum_is_visible = true;
+    return;
+  }
+
   def_width = PUM_DEF_WIDTH;
   max_width = 0;
   kind_width = 0;
   extra_width = 0;
 
   // Pretend the pum is already there to avoid that must_redraw is set when
   // 'cuc' is on.
```

Comment:
Actually I think we could stop saying that we're pretending and just let `pum_is_visible` be true from here on.